actions/analyze_rosbag_image_exposure/src/analyze_rosbag_image_exposure/__main__.py

```diff
@@ -0,0 +1,485 @@
+
+    # enumerate all image paths
+    for bag_folder in bag_out_folders:
+        bag_put_tags = []
+        all_img_paths = []
+        for img_ext in ["jpg", "JPG", "JPEG", "png", "PNG"]:
+            all_img_paths.extend(glob(f"**/*.{img_ext}", root_dir=str(bag_folder), recursive=True))
+
+        if len(all_img_paths) == 0:
+            logger.warning(f"No image files extracted from {bag_folder.name}.bag.")
+    
+        overexposed_images = []
+        underexposed_images = []
+        good_images = []
+        for img_path in all_img_paths:
+
+            # read in image in grayscale mode
+            img = cv2.imread(str(bag_folder / img_path), cv2.IMREAD_GRAYSCALE)
+            
+            # take a histogram of the image. By default we take 15 bins
+            hist = np.histogram(img, bins=bins, range=(0., 255.))[0]
+            largest_bin = np.argmax(hist)
+            
+            # add a file-level tag if image is over or underexposed
+            # images with most intensity at the bottom of the histogram are underexposed
+            if largest_bin == 0:
+                put_tag = "underexposed"
+                underexposed_images.append(img_path)
+            # images with most intensity at the top of the histogram are overexposed
+            elif largest_bin == bins - 1:
+                put_tag = "overexposed"
+                overexposed_images.append(img_path)
+            else:
+                put_tag = False
+                good_images.append(img_path)
+            
+            if put_tag:
+                bag_put_tags.append(put_tag)
+                if dataset and extract_all_images:
+                    print(f"putting tag {put_tag} on imag {img_path}")
+                    FileManager.put_tags(f"{bag_folder.name}/{img_path}", [put_tag])
+
+
+        # put tags on rosbag if dataset is not None and put_tags exist
+        if dataset: 
+            if bag_put_tags:
+                #bag_fname = pathlib.Path(input_file_or_folder) / f"{bag_folder.name}.bag"
+                bag_fname = f"{bag_folder.name}.bag"
+                # build a MetadataChangeset with put_tag direction
+                changeset = updates.MetadataChangeset().Builder()
+                # put tags on the Changeset builder one at a time
+                for tag in set(bag_put_tags): # only tag dataset with each tag value once
+                    changeset = changeset.put_tag(tag)
+                changeset = changeset.build()
+                bag_file = dataset.get_file_info(bag_fname)
+                file_update_request = files.file_requests.UpdateFileRecordRequest(metadata_changeset = changeset)
+                # pass request to add metadata
+                bag_file = bag_file.update(file_update_request)
+                logger.info(f"Tagging {bag_fname} with {bag_put_tags}")
+
+        # save some example images, max 5 for each of over and underexposed
+        max_overex_examples = min(len(overexposed_images), 5)
```

Comment:
would be nice to parameterize the number of images to save

Reply:
separately from `extract_all`?